### PR TITLE
[top] Generate reset request headers

### PR DESCRIPTION
**Pull request by tjaychen**
Signed-off-by: Timothy Chen <timothytim@google.com>

### CODE REVIEW

**Comment by tjaychen** [comment]
@lenary please have a look.
This added enum eventually is meant to apply to both `rstmgr` and `pwrmgr`, so i'm not sure if i should name it differently?
I really want to name it as `top_earlgrey_reset_requests`, but i'm not sure if that's best. 

@alphan  FYI
@silvestrst  FYI

**Comment by silvestrst** [comment]
@tjaychen thank you for "CC" me.

To better understand what the naming should be, could you please describe how this enum is intended to be used in Power Manager and Reset Manager?

**Comment by lenary** [comment]
If we're going to share reset request reasons between power manager and reset manager, remove "power manager" from the name, i think - though we should ensure the IDs are useful for both sides.

I presume on the power manager side this is for indexing into `PWRMGR.RESET_EN` and `PWRMGR.RESET_STATUS`, and maybe also `PWRMGR.WAKE_INFO.REASONS`. I cannot see where this is intended to be used on the reset manager side.

**Comment by tjaychen** [comment]
if you see the latest reset manager spec (just merged!) there is a section
on reset reasons.
The reset reason is actually maintained in the reset manager, even though
the reset request enables are maintained in the power manager.

I actually think long term i should migrate the latter to the former. In
either case, the encoding and bit-position of both of these will be
consistent and the same.

**Comment by tjaychen** [comment]
thinking about it more, I really do want to move the reset request connections to rstmgr. This makes the whole thing cleaner and more contained. But I will probably do that as a future PR. For now, I think it makes sense to just call this `pwrmgr` reset requests. I can then modify in the future.

Let me know what you guys think.

**Comment by tjaychen** [comment]
@silvestrst let me know what you prefer. 
I can try to move the reset request function from `pwrmgr` to `rstmgr` before merging this PR.
Then the `rstmgr` reg interface will be more or less complete.

Or we can go with this first, and do it later.

The main difference is that the reset request enable registers will move from `pwrmgr` to `rstmgr`

@alphan FYI

File: hw/top_earlgrey/sw/autogen/top_earlgrey.h
Comment (alphan) [inline hw/top_earlgrey/sw/autogen/top_earlgrey.h:779]: If this is going to apply to both power manager and reset manager, it is probably a good idea to remove "power manager" from the name, as @lenary suggests.

**Comment by alphan** [comment]
> @silvestrst let me know what you prefer.
> I can try to move the reset request function from `pwrmgr` to `rstmgr` before merging this PR.
> Then the `rstmgr` reg interface will be more or less complete.
> 
> Or we can go with this first, and do it later.
> 
> The main difference is that the reset request enable registers will move from `pwrmgr` to `rstmgr`
> 
> @alphan FYI

I saw your comment after reviewing the change. It looks like moving the reset request function to `rstmgr` will be a bigger change and will also require changes in the pwrmgr DIF library (#3142, #3640) but I'm all for it if it will result in a simpler/cleaner design. If you choose to do it in a separate PR, then I guess you can keep the current names and update them in that PR.

I remember we briefly talked about this but can you remind me why reset request enables were handled by the pwrmgr in the first place? I thought reset requests needed some coordination in the pwrmgr side.

**Comment by tjaychen** [comment]
> > @silvestrst let me know what you prefer.
> > I can try to move the reset request function from `pwrmgr` to `rstmgr` before merging this PR.
> > Then the `rstmgr` reg interface will be more or less complete.
> > Or we can go with this first, and do it later.
> > The main difference is that the reset request enable registers will move from `pwrmgr` to `rstmgr`
> > @alphan FYI
> 
> I saw your comment after reviewing the change. It looks like moving the reset request function to `rstmgr` will be a bigger change and will also require changes in the pwrmgr DIF library (#3142, #3640) but I'm all for it if it will result in a simpler/cleaner design. If you choose to do it in a separate PR, then I guess you can keep the current names and update them in that PR.
> 
> I remember we briefly talked about this but can you remind me why reset request enables were handled by the pwrmgr in the first place? I thought reset requests needed some coordination in the pwrmgr side.

yes that's true, the reset requests do need coordination with `pwrmgr`.  However, `pwrmgr` doesn't actually need to know "which" reset request, just that there was "a request".  So I think ideally the way this should work is..

reset requests -> rstmgr (contains enable logic) -> forwards OR'd reset request to pwrmgr for coordination -> reset happens / rstmgr records reset cause.

Instead, the way it works right now is..

reset requests -> pwrmgr (contains enable logic) -> reset happens -> pwrmgr informs rstmgr which reset to record.

So they both work, but I think by changing it is more consistent with what each block is doing.

**Comment by silvestrst** [comment]
Hi @tjaychen ,

in `rstmgr` we have `RSTMGR.SW_RST_REGEN` and `RSTMGR.SW_RST_CTRL_N` for the peripherals. In `pwrmgr` we have `PWRMGR.RESET_EN_REGWEN`, `PWRMGR.RESET_EN`. Are you proposing merging the functionality of these `pwrmgr` registers into the `RSTMGR.SW_RST_REGEN` and `RSTMGR.SW_RST_CTRL_N`?

Am I understanding correctly that you are proposing moving all reset requests and enable to the `rstmgr` side, whilst now there are some in `rstmgr` and some in `pwrmgr`? If that is the case then it makes sense to me.

I don't mind in what sequence the changes are made, if you make the proposed HW changes I will just catch up on the SW side when you are done.

Thanks!

**Comment by tjaychen** [comment]
> Hi @tjaychen ,
> 
> in `rstmgr` we have `RSTMGR.SW_RST_REGEN` and `RSTMGR.SW_RST_CTRL_N` for the peripherals. In `pwrmgr` we have `PWRMGR.RESET_EN_REGWEN`, `PWRMGR.RESET_EN`. Are you proposing merging the functionality of these `pwrmgr` registers into the `RSTMGR.SW_RST_REGEN` and `RSTMGR.SW_RST_CTRL_N`?
> 
> Am I understanding correctly that you are proposing moving all reset requests and enable to the `rstmgr` side, whilst now there are some in `rstmgr` and some in `pwrmgr`? If that is the case then it makes sense to me.
> 
> I don't mind in what sequence the changes are made, if you make the proposed HW changes I will just catch up on the SW side when you are done.
> 
> Thanks!

sorry for the late reply. 

The `SW_RESET_REGEN` and `SW_RESET_CTRL_N` fields are specifically for peripherals with software reset capability (there are only 2 at the moment).  So these basically control blocks that software can manually reset if it chooses to.

The `RESET_EN` in `pwrmgr` on the other hand, controls which reset request from the peripherals will be honored.  Certain peripherals may request the system to reset (for example a timer watchdog), so that's the function these bits are controlling.  These functions really should be moved to `rstmgr` imo and not left in `pwrmgr`.  The power manager doesn't really care who is requesting the reset, only that someone has.

**Comment by silvestrst** [comment]
> The RESET_EN in pwrmgr on the other hand, controls which reset request from the peripherals will be honored. Certain peripherals may request the system to reset (for example a timer watchdog), so that's the function these bits are controlling. 

Thank you for the explanation!

**Comment by lenary** [comment]
@tjaychen what's the status of this? Is it blocked on a hardware change (moving resets into power manager), or something else?

**Comment by tjaychen** [comment]
no it's not blocked.
Just when the hardware gets an update, this will have to be updated also.

I'm guessing you guys are okay with that for now? I can rebase and merge.

**Comment by silvestrst** [comment]
I think there might be unresolved comment, but otherwise I am happy. 

> Just when the hardware gets an update, this will have to be updated also.

Wouldn't it automatically pick up the HW changes (which I guess will be just extra reset signals)?

**Review by alphan — APPROVED** [review]
Apologies for the delay, LGTM. Please feel free to ignore my comment regarding naming if you want to address it later. Can you also ping @silvestrst and me or create an issue when the changes that you mentioned land so that we can update the DIFs?

**Comment by tjaychen** [comment]
sorry when i said hardware changes, i'm refering mainly to the fact that
reset enable controls will migrate from pwrmgr to rstmgr.
So in that scenario, we will need to at least make some small changes to
the DIFs.

**Pull request merged**
